flow/raster_cache.h

```diff
@@ -42,6 +53,44 @@ class RasterCacheResult {
 
 struct PrerollContext;
 
+// The |RasterCacheImageDelegate| creates and retruns a cached result of
+// rendering the supplied |Layer| or |SkPicture|.
+class RasterCacheImageDelegate {
+ public:
+  virtual std::unique_ptr<RasterCacheResult> RasterizePicture(
+      SkPicture* picture,
+      GrContext* context,
+      const SkMatrix& ctm,
+      SkColorSpace* dst_color_space,
+      bool checkerboard) const = 0;
+
+  virtual std::unique_ptr<RasterCacheResult> RasterizeLayer(
+      PrerollContext* context,
+      Layer* layer,
+      const SkMatrix& ctm,
+      bool checkerboard) const = 0;
+};
+
+// The |SkRasterCacheImageDelegate| creates and returns rendered results
+// for |Layer| and |SkPicture| objects stored as |SkImage| objects.
+class SkRasterCacheImageDelegate : public RasterCacheImageDelegate {
+ public:
+  std::unique_ptr<RasterCacheResult> RasterizePicture(
+      SkPicture* picture,
+      GrContext* context,
+      const SkMatrix& ctm,
+      SkColorSpace* dst_color_space,
+      bool checkerboard) const override;
+
+  std::unique_ptr<RasterCacheResult> RasterizeLayer(
+      PrerollContext* context,
+      Layer* layer,
+      const SkMatrix& ctm,
+      bool checkerboard) const override;
+
+  static SkRasterCacheImageDelegate instance;
```

Comment:
For a singleton this would be better as a function instead of a direct variable.  That way you can swap out the implementation and introduce lazy evaluation etc.

Reply:
A singleton pattern would also make constructors private, and have a static public `getInstance` function as the only way to retreive the singleton. For simplicity, I think it's probably easier to just remove this delegate class and have `RasterCache`, `MockRasterCache` to directly specify the `RasterizePicture` and `RasterizeLayer` behaviors.

Reply:
Class eliminated...